pkg/virt-controller/watch/vmi.go

```diff
@@ -477,158 +480,46 @@ func (c *VMIController) updateStatus(vmi *virtv1.VirtualMachineInstance, pod *k8
+			if container.Name == "compute" {
+				foundImage = container.Image
+				break
 			}
-
-			patchOps = append(patchOps, fmt.Sprintf(`{ "op": "test", "path": "/status/activePods", "value": %s }`, string(oldPods)))
-			patchOps = append(patchOps, fmt.Sprintf(`{ "op": "replace", "path": "/status/activePods", "value": %s }`, string(newPods)))
-
-			log.Log.V(3).Object(vmi).Infof("Patching VMI activePods")
 		}
+		vmiCopy = c.setLauncherContainerInfo(vmiCopy, foundImage)
 
-		if vmiPodExists {
-			var foundImage string
-
-			for _, container := range pod.Spec.Containers {
-				if container.Name == "compute" {
-					foundImage = container.Image
-					break
-				}
-			}
-
-			vmiCopy = c.setLauncherContainerInfo(vmiCopy, foundImage)
-
-			if vmiCopy.Status.LauncherContainerImageVersion != vmi.Status.LauncherContainerImageVersion {
-				if vmi.Status.LauncherContainerImageVersion == "" {
-					patchOps = append(patchOps, fmt.Sprintf(`{ "op": "add", "path": "/status/launcherContainerImageVersion", "value": "%s" }`, vmiCopy.Status.LauncherContainerImageVersion))
-				} else {
-					patchOps = append(patchOps, fmt.Sprintf(`{ "op": "test", "path": "/status/launcherContainerImageVersion", "value": "%s" }`, vmi.Status.LauncherContainerImageVersion))
-					patchOps = append(patchOps, fmt.Sprintf(`{ "op": "replace", "path": "/status/launcherContainerImageVersion", "value": "%s" }`, vmiCopy.Status.LauncherContainerImageVersion))
-				}
-			}
-
-			if !reflect.DeepEqual(vmi.Labels, vmiCopy.Labels) {
-				labelBytes, err := json.Marshal(vmiCopy.Labels)
-				if err != nil {
-					return err
-				}
-				origLabelBytes, err := json.Marshal(vmi.Labels)
-				if err != nil {
-					return err
-				}
-
-				if vmi.Labels == nil {
-					patchOps = append(patchOps, fmt.Sprintf(`{ "op": "add", "path": "/metadata/labels", "value": %s }`, string(labelBytes)))
-				} else {
-					patchOps = append(patchOps, fmt.Sprintf(`{ "op": "test", "path": "/metadata/labels", "value": %s }`, string(origLabelBytes)))
-					patchOps = append(patchOps, fmt.Sprintf(`{ "op": "replace", "path": "/metadata/labels", "value": %s }`, string(labelBytes)))
+	case vmi.IsScheduled():
+		// Nothing here
+		break
+	default:
+		return fmt.Errorf("unknown vmi phase %v", vmi.Status.Phase)
+	}
 
-				}
-			}
+	// VMI is owned by virt-controller, so patch instead of update
+	if vmi.IsRunning() || vmi.IsScheduled() {
+		patchBytes, err := preparePatch(vmi, vmiCopy)
+		if err != nil {
+			return fmt.Errorf("error preparing VMI patch: %v", err)
 		}
 
-		if len(patchOps) > 0 {
-			patch := "[ "
-			for i, entry := range patchOps {
-				patch += entry
-
-				if i == len(patchOps)-1 {
-					patch += " ]"
-				} else {
-					patch += ", "
-				}
-			}
-
-			_, err = c.clientset.VirtualMachineInstance(vmi.Namespace).Patch(vmi.Name, types.JSONPatchType, []byte(patch))
+		if len(patchBytes) > 0 {
+			_, err = c.clientset.VirtualMachineInstance(vmi.Namespace).Patch(vmi.Name, types.JSONPatchType, []byte(patchBytes))
```

Comment:
why was this moved outside of the switch case? This is only relevant when the vmi is in the Running phase.

Reply:
This was initially done since the updating of the Ready condition now occurs in multiple phases, and not only when Running as before (i.e., the `syncReadyConditionFromPod` method is called outside of the switch case block). When testing locally, it appeared to me that virt-controller/virt-handler were racing in updating the VMI (both logs showed many occurences of "_the object has been modified; please apply your changes to the latest version and try again_"). In practice I'm not convinced whether it's possible for the Ready condition to be updated when in Scheduled phase or not, but I figured it would anyway be useful to decouple from the logic that updates the local state of the object (the switch case), and the logic that determines how to update the remote object (patch vs. update).

Reply:
@davidvossel I think this remains the only unresolved discussion in this PR. Please see my above comment, and let me know if you prefer me to undo that change. Thanks! :-)

Reply:
i think it's fine to keep it as you have it now